[APP] SignUpRIB을 구현했어요.

allie0147 opened the pull request:
- Resolved #39 
---

### 설명
- `SignUpRIB`을 구현했어요.
- SignUp 화면을 구성했어요.
- `DesignSystem`에 이미지를 추가했어요.
- `SignUpRequestModel`를 만들었어요.

### 작업사항
- `Dependency` 추가 : `RxAppState`, `RxGesture`
- `SignUpRIB` 추가 :  `SignInRIB`에 `SignUpRIB` 연결
- 화면 구성 : `UIPageViewController`에 `subView` 4개 추가
- 이미지 추가 : 버튼 아이콘, 회원가입 이미지 추가
- `Enum` 생성 : `SignUpState`, `Gender`, `VeganState`
- 버튼 생성 : `ImageButton`
- `Notification` 추가 : `BottomButton` 활성화/비활성화 이벤트 전송
- `SignUpRequestModel` 생성

### 스크린샷

https://user-images.githubusercontent.com/62925639/180445902-99d7d23a-3d3f-4954-adcc-bdd5c9342af2.MP4


### 기타사항
`SignUpRIB`에서 뒤로가기 했을때 `SignInRIB`으로 가야하는데 연결을 어떻게 해야할지 모르겠어요 ;;
`SignInRIB`이 `SplashRIB`과 연결되어 있어서 `Splash`에서 분리해야 할 것 같아요.

생각보다 할 기능이 많아서 늦어졌네요 ㅠㅠ 
최대한 나눠서 commit 해봤습니다만,, 사용한 코드가 많아서 commit이 많아졌어요 ..
피드백 환영입니다 !!!

allie0147 committed:
Feat : SignUp RIB 추가

allie0147 committed:
Feat : RxAppState 추가

allie0147 committed:
Feat : VegannerApp에 RxAppState 추가

allie0147 committed:
Feat: SignUp UI 구성

allie0147 committed:
Refactor : `SplashRIB`에 `SignUpRIB` 연결

allie0147 committed:
Fix : `SplashRIB` 다시 원복

allie0147 committed:
Feat : 회원가입에 사용될 이미지 추가

allie0147 committed:
Feat : 아이콘 추가

allie0147 committed:
Fix : Button들 event 수정

allie0147 committed:
Feat : `RxGesture` 추가

allie0147 committed:
Feat : SignUp에서만 쓰일 `ImageButton` 추가

allie0147 committed:
Feat : `SignUpRIB` 구성

allie0147 committed:
Feat : `SignUpPageViewController` 구성

allie0147 committed:
Feat : SignUp 첫번째 뷰 구성

allie0147 committed:
Feat : SignUp 두번째 뷰 구성

allie0147 committed:
Feat : SignUp 세번째 뷰 구성

allie0147 committed:
Feat : SignUp 마지막 뷰 구성

allie0147 committed:
Feat : SignUp에 쓰일 enum들 추가

allie0147 committed:
Feat : `SignInRIB`에 `SignUpRIB` 연결

allie0147 committed:
Feat : BottomButton에 쓰일 Notification 추가

allie0147 committed:
Feat : `SignUpRequestModel` 추가

allie0147 committed:
Feat : request model 만들기

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SignUpViewController.swift:
```diff
@@ -0,0 +1,127 @@
+
+        guard pageViews.count > previousIndex else { return nil }
+
+        return pageViews[previousIndex]
```
배열조회할때 해당 Extension을 적용해서 가져오는 방법도 있어요
http://minsone.github.io/programming/check-index-of-array-in-swift

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SignUpViewController.swift:
```diff
@@ -6,122 +6,232 @@
+    case two
+    case three
+    case four
+}
```
스텝이름이 숫자로 되어있으면 확장성이 없을거 같아요 그리고 파일 이름으로 해당 화면이 어떤화면인지 유추하기도 쉽지 않을거라 생각해요
예를들면 닉네임 받는 화면과 성별 받는 화면의 순서를 변경해달라고 기획에서 요청들어왔을때
숫자로 되어있으면 변경해야하는 공수가 클거라고 생각해요

```
enum SignUpSteps {
case nickname
case gender
case birthday
}
```

저는 이렇게 표현하는것이 좋다고 생각해요

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SignUpViewController.swift:
```diff
@@ -6,122 +6,232 @@
+            .bind(onNext: {
+            self.configureLayout()
+        }).disposed(by: disposeBag)
+
```
해당 로직은 viewDidLoad 시점에 적용해도 될 거 같은데
viewDidLayoutSubview 후 take(1) 하는 이유가 있나요??

allie0147 replied:
`viewDidLoad`에서는 `view.safeAreaInsets.bottom`을 가져오지 못하기 떄문에 `didLayoutSubViews`에서 적용했어요.
`UIApplication.shared.keyWindow`로 값을 가져올 수 있지만,
 iOS 13부터 `deprecated` 되어서 `didLayoutSubViews` 에 적용했습니다.
그리고 UI가 update될때마다 `didLayoutSubViews`가 불리기 떄문에 `.take(1)`을 하지 않게되면 
`self.configureLayout()`이 계속 그려져서 `bottomButton` 높이가 계속 변경되기 떄문이에요.

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SignUpViewController.swift:
```diff
@@ -6,122 +6,232 @@
+            stepIconImageView.image = UIImage(asset: VegannerAppDevAsset.iconStep04)
+            stepProgressBar.setProgress(1.0, animated: true)
+            stepProgressLabel.text = "100%"
+        }
```
enum 값 내에 변수를 만들어서 사용하는 방법이 유지보수와 가독성 측면에 이점이 있을것으로 생각돼요
```
enum Steps {
case aa 
case bb
case cc
}

extension Steps {
var title: String {
switch self {
case aa: return "step 1"
case bb: return "step 2"
case cc: return "step 3"
}
}
}
....
naviagationTitle.text = step.title
```

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SubViews/SignUpFirstStepVIewController.swift:
```diff
@@ -7,13 +7,77 @@
+            } else {
+                self?.nicknameTextField.isValidText = false
+                NotificationCenter.default.post(name: .bottomButtonState, object: false)
+            }
```
저는 해당 처리는 비즈니스로직으로 보이는데 
View보다는 Interactor에서 처리하는게 맞다고 생각이 듭니다

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SubViews/SignUpFirstStepVIewController.swift:
```diff
@@ -7,13 +7,77 @@
+    private func isValidNickname(_ nickname: String) -> Bool {
+        let nicknameRegex = "^[ㄱ-ㅎ|가-힣|a-z|A-Z|0-9]+$"
+        let nicknameTest = NSPredicate(format: "SELF MATCHES %@", nicknameRegex)
+        return nicknameTest.evaluate(with: nickname)
```
이 로직도 같이 interactor로 가게되겠네요

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SubViews/SignUpPageViewController.swift:
```diff
@@ -0,0 +1,107 @@
+            SignUpFirstStepViewController(),
+            SignUpSecondStepViewController(),
+            SignUpThirdStepViewController(),
+            SignUpLastStepViewController()
```
PageViewController 내 ChildViewController를 선언하게 되면
PageViewController와 ChildViewController와의 강한 의존성을 갖게 되는데
빌더에서 주입받는 형태가 좋지 않을까요??

MOON-HJ commented on SixthSense/VegannerApp/Sources/SignUp/SubViews/SignUpThirdStepViewController.swift:
```diff
@@ -7,13 +7,107 @@
+            .distinctUntilChanged()
+            .bind { [weak self] text in
+            print("text~ \(text)")
+            NotificationCenter.default.post(name: .bottomButtonState, object: true)
```
디버그 프린트 문 제거 부탁드릴게요

그리고 Interaction에서 따로 버튼에 대한 상태를 바인딩하지 않고 NotificationCenter로 처리하신 이유가 있나요??

allie0147 replied:
`bottom button state`가 **subVC -> pageVC -> signupVC**로 이동해야 하기 떄문에 `notification`을 사용했어요.
그리고 `bottom button`으로 `pageVC`를 이동시켜야하기때문에 최상단인 `signupVC`에 `button`을 넣었습니다.

MOON-HJ replied:
notification자체가 앱 내 전체적으로 이벤트를 방출시킬 수 있어 나중에 오류 발생시 디버깅하기 힘들거 같아요
그리고 저는 해당 로직은 interactor에서 처리해야한다고 보는데 View에서 처리하시는 이유가 있나요??

MOON-HJ commented on SixthSense/DesignSystem/Sources/Buttons/SelectButton.swift:
```diff
@@ -18,7 +18,7 @@ public final class SelectButton: UIButton, ButtonProtocol {
         didSet {
             let title = titleLabel?.text ?? ""
-            hasFocused ? setButtonTitle(with: title) : setButtonTitleFocused(with: title)
+            hasFocused ? setButtonTitleFocused(with: title) : setButtonTitle(with: title)
```
추후에 프로퍼티 옵저버로 컨트롤하는 로직들은 메소드로 빼셨으면 좋겠어요
프로퍼티옵저버 내 로직이 길어지면 코드리딩하기 힘들어요ㅠ

MOON-HJ commented on SixthSense/Account/Sources/SignUp/Presentation/SubViews/SignUpSecondStepViewController.swift:
```diff
@@ -0,0 +1,161 @@
+        stack.spacing = 12
+    }
+
+    private let maleButton = SelectButton(title: "남성").then {
+        $0.tag = 0
+    }
+    private let femaleButton = SelectButton(title: "여성").then {
+        $0.tag = 1
+    }
+    private let etcButton = SelectButton(title: "그 외 성별").then {
+        $0.tag = 2
+    }
+    private let noneButton = SelectButton(title: "선택 안 함").then {
+        $0.tag = 3
+    }
```
Button 누르는 로직을 한번에 사용하시려고 tag를 사용하신건가요??
그렇다면 enum값으로 관리부탁드립니다

MOON-HJ commented on SixthSense/Account/Sources/SignUp/Presentation/SubViews/SignUpThirdStepViewController.swift:
```diff
@@ -0,0 +1,132 @@
+            .debug()
+            .bind { isNotEmpty in
+            NotificationCenter.default.post(name: .bottomButtonState,
+                                            object: isNotEmpty)
```
해당 비즈니스 로직도 interactor에서 처리해야할 것으로 보이는데 수정부탁드려요

MOON-HJ commented on SixthSense/Account/Sources/SignUp/Presentation/SignUpViewController.swift:
```diff
@@ -228,6 +231,23 @@ private extension SignUpViewController {
         }).disposed(by: disposeBag)
     }
 
+    private func applyData(_ data: [SignUpStep: String]) {
+        guard let step = data.keys.first else { return }
+        guard let value = data.values.first else { return }
+
+        switch step {
+        case .exit: return
+        case .one:
+            signUpRequestModel.nickName = value
+        case .two:
+            signUpRequestModel.gender = value
+        case .three:
+            signUpRequestModel.birthDay = value
+        case .four:
+            signUpRequestModel.vegannerStage = value
+        }
+    }
+
```
signUpRequestModel로 데이터 관리하는 처리는 `SignUpInteractor`에서 처리해야할 거 같아요

MOON-HJ commented:
> SignUpRIB에서 뒤로가기 했을때 SignInRIB으로 가야하는데 연결을 어떻게 해야할지 모르겠어요 ;;

Listener에 `detachSignUp()` 추가하시고 SignInRouter에서 detachSignUp에다가 dismiss 처리하면 될 거 같아요
그리고 SignUpInteractor에서 뒤로가기 버튼 눌렸을때 detach 호출하면 될 거 같아요

MOON-HJ reviewed (changes requested):
전반적으로 뷰에 비즈니스 로직처리가 많이 들어가있네요ㅠㅠ
해당 로직들을 인터렉터 내부로 옮겨주시고
그리고 저희 룰로 가져가고 싶은게 10커밋 10파일체인지(리소스제외)로 가져갔으면 좋겠어요

이번 PR도 SignUp하나, 리소스하나, 스텝 각각해서 6개로 분리할 수 있을거 같은데
한 PR로 올라오니 PR 확인하는데 집중력이 많이 떨어지네요ㅠ

MOON-HJ committed:
Interactor 피드백 반영 및 수정

Fix : `Image.assets` -> `Assets.assets` 변경

Feat : Array Extension 추가

Fix : Enum에 extension 추가

Fix : 변경된 enum 적용, subview 이름 변경

Fix : 빌드 오륲 수정

Fix : GenderView 로직 interactor로 수정

Fix : GenderView button state optional 제거

- 초기값이 없어서 optional일 필요가 없음

Fix : GenderView interactor relay 수정

Fix : BirthView 로직 interactor로 수정

Fix : VeganStageView 로직 interactor로 수정

Fix : dataModel interactor로 이동

Fix : handler로직 수정

MOON-HJ commented on SixthSense/Account/Sources/SignUp/Presentation/SubViews/SignUpPageViewController.swift:
```diff
@@ -13,15 +13,15 @@ import UIKit
+            GenderStepViewController(),
+            BirthStepViewController(),
+            VeganStepViewController()
         ])
```
해당 VC에 있는 로직들도 Interactor로 옮겨주셨으면 좋겠습니다

allie0147 commented:
https://github.com/TheSixthSense/app-ios/pull/48/commits/d5f07f6a69726c45fda1c859caeae391e47083c5
@MOON-HJ  수정반영본 확인해주세요!

MOON-HJ reviewed (approved):
👍🏼 피드백 반영하시느라 고생많으셨습니다 👍🏼
